### PR TITLE
MM-53237 - Removefile count in the RHS channel info panel - patch for v7.X

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
This PR places the file count behind a config setting found in system console > file storage > Show Channel File Count. The reason behind this is the way the query is currently structured and that its causing DB performance issues to certain customers which have in their workspace channels with high volume of files. The idea behind this PR is to give the option to those customers to have the ability to disable the file count preventing this DB outages.

Also, it was found during the implementation of the solution that there were two other places that were requesting the channel stats including the fileCount which was completely unnecessary for those cases. Those are:

- when adding/removing users from a channel.
- when editing a channel in the system console.

In the attached videos it is shown how while performing any of the two previous mentioned actions the file count was being pulled unnecessarily. After the changes applied by this PR, only the required data for the action is fetched avoiding pulling the unnecessary fileCount data.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53237

#### Screenshots

Before:

https://github.com/mattermost/mattermost/assets/10082627/53872ed8-6454-48e5-a7c3-1ed463dc0af4

https://github.com/mattermost/mattermost/assets/10082627/74163542-9566-4a66-a851-a0476180763c

https://github.com/mattermost/mattermost/assets/10082627/55015eb7-8df2-477b-8290-8f8c6b6963bb

After:

https://github.com/mattermost/mattermost/assets/10082627/a21ec76d-69ca-4986-a272-909a94aa52d3

https://github.com/mattermost/mattermost/assets/10082627/d989f1e1-9e90-4eba-9eb1-7a40a08ee6a8

#### Release Note
```release-note
Add a new configuration option to show/hide the Channel File Count shown in the RHS channel info panel. The Channel File Count will be enabled by default and can be configured via the System Console > File Storage > Show Channel File Count.
```

### CODE REVIEW

File: webapp/platform/client/src/client4.ts
Comment (pvev) [inline webapp/platform/client/src/client4.ts:1750]: This change prevents that a stats request from any other place different from the Channel Info RHS panel would accidentaly include the file count request. This is the right way since the file count is only used there and now will be conditioned by the config value.

File: webapp/platform/components/src/generic_modal/__snapshots__/generic_modal.test.tsx.snap
Comment (pvev) [inline webapp/platform/components/src/generic_modal/__snapshots__/generic_modal.test.tsx.snap:3]: while running my tests locally it had to update this snapshot since it was outdated.

**Comment by matthewbirtch** [comment]
> @matthewbirtch this is the solution we have come with for the described problem. (TL;DR the fileCount value shown in the channel info RHS panel is causing performance problems to customers with high volume of files). I wanted to let you know in case that from UX you have an strong opinion about the provided value of this fileCount data.

To be honest @pvev I'm not certain the file count is valuable enough to add a config setting for it. If it's causing significant performance issues, did we consider just removing the count altogether? I doubt people would miss it or even notice it that much to be honest. 

cc: @johndavidlugtu @abhijit-singh as this is their area of ownership I believe

**Comment by johndavidlugtu** [comment]
Good stuff here @pvev -- thanks for looping me in @matthewbirtch 

I'm +1 that there isn't much value in knowing how many files are in a channel, but I'm open to hearing reasons otherwise (@abhijit-singh if you have thoughts)

A question and a comment:
1. _If_ we opt to keep the file count, can we run it once they click into "File" and show the number there instead? That way it isn't called immediately. Thoughts?
Screenshot 2023-06-21 at 2 04 33 PM

2. _If_ we opt to give the option to disable file count, I think the path here would be to disable it by default for everyone.

**Comment by pvev** [comment]
@johndavidlugtu answering your questions:

> 1. _If_ we opt to keep the file count, can we run it once they click into "File" and show the number there instead? That way it isn't called immediately. Thoughts?
It is going to definitely reduce the amount of calls to this query, but for the affected customers will still be painful even to run it just once, they will still notice the performance degradation for that scenario while clicking the files section.


> 2. _If_ we opt to give the option to disable file count, I think the path here would be to disable it by default for everyone.
> Also, just a heads up for visibility @marianunez
Agree. 

Also, if there is not much significant value in seeing the fileCount, we should consider @matthewbirtch suggestion of removing it altogether since this has proven to be causing too much trouble vs. the potential value provided by knowing the num of existing files per channel.

There was already an effort to improve that query but the problem is the migration process and how error prone, expensive in time and painful would it be for customers with high files volume to perform that migration. Probably not worth it.

My personal opinion is that the config setting is a valid option and is ready to go with this PR once the reviews are complete,  and we can send it disabled by default and if for any reason a customer wants it or ask for it, they can always enable it under their own risk.

**Comment by matthewbirtch** [comment]
> My personal opinion is that the config setting is a valid option and is ready to go with this PR once the reviews are complete, and we can send it disabled by default and if for any reason a customer wants it or ask for it, they can always enable it under their own risk.

@pvev @johndavidlugtu I think we need to be more mindful of how many config settings we have. In order for us to add yet another config setting in the system console, there needs to be a threshold of value in my opinion. Otherwise we run the risk of providing too many settings to digest and navigate through. I don't believe this count provides enough value to warrant a setting and I would be more in favour of removing the count altogether if it's causing too many performance issues.

**Comment by marianunez** [comment]
> @pvev @johndavidlugtu I think we need to be more mindful of how many config settings we have. In order for us to add yet another config setting in the system console, there needs to be a threshold of value in my opinion. Otherwise we run the risk of providing too many settings to digest and navigate through. I don't believe this count provides enough value to warrant a setting and I would be more in favour of removing the count altogether if it's causing too many performance issues.

+1 on simplifying the solution and avoiding extra config settings for this minor detail of File counts. We should overall remove it and rethink on where and how we could bring it back - an option as John mentioned could be once you view the files in the RHS

**Comment by johndavidlugtu** [comment]
Thanks everyone for the input! @pvev let's remove file count and not provide a config setting.

> We should overall remove it and rethink on where and how we could bring it back - an option as John mentioned could be once you view the files in the RHS

+1

**Comment by pvev** [comment]
> Thanks everyone for the input! @pvev let's remove file count and not provide a config setting.
> +1
ok, thank you all for the constructive analysis and discussion. I will proceed to update this PR to do an initial and quick removal of the feature from the front-end and open a new ticket to handle the remove of all the backend stuff including the query itself.
cc: @Willyfrog

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking review until the PR has been updated

**Review by matthewbirtch — APPROVED** [review]
Thanks @pvev!

File: webapp/platform/client/src/client4.ts
Comment (hanzei) [reply webapp/platform/client/src/client4.ts:1750]: :+1: for this change. The query is expensive and hence should be opt-in.

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for driving this long discussion @pvev :+1: 

I'm not quite convinced that removing the file count is necessary. Please see my comment on the original discussion in MM: https://community.mattermost.com/private-core/pl/fgu3938eff8y9btoqdscxqj57w

**Review by JulienTant — APPROVED** [review]
Looks good @pvev

**Comment by pvev** [comment]
heads up that for v8.0 there was already a [change in place ](https://github.com/mattermost/mattermost/pull/22603) to fix the fileCount query so this PR is no longer necessary for v8.0 @johndavidlugtu @matthewbirtch @marianunez . However, I wonder if we should patch 7.X with the proposed change in this PR since the fix for the query should not be used as patch. Thoughts? @Willyfrog

**Comment by Willyfrog** [comment]
Can the patch go behind a feature flag or a system console setting? 

I don't like the idea of doing a patch for only a dot release as it might cause confusion, but I'm only 1/5 on this (easily convinced ;) )

**Review by hanzei — COMMENTED** [review]
Will the config value `ShowChannelFileCount` still be part of 8.0? Do we need it moving forward?

On a broader scope: given that 8.0 is out, do we need to fix the performance issue for 7.X?

**Comment by grubbins** [comment]
Is this still planned for release as a patch? This is still affecting us a lot...

**Comment by hanzei** [comment]
@grubbins Did you update your server to v8.0? There are some performance fixes in that release that might help with your issue.

**Comment by grubbins** [comment]
Well, we're unlikely to go to 8.x immediately for stability reasons - we
were operating only on ESRs until recently (they seem to have stopped
coming out?). Probably planning to wait for a stability release if they
still happen, or at least for 8.1.

We went to 7.10 somewhat against the above policy on the selling point of
various other performance improvements - if we could get this in 7.10
sooner, I'd be very happy about it! We have to rent a considerably larger
DB instance in AWS than we think we need at the moment, to handle some of
these queries.

**Comment by hanzei** [comment]
The next ESR will be v8.1, which will be released on the 16th of August. Is it reasonable to ask you to update once the next ESR comes out in a week?

**Comment by grubbins** [comment]
IIUC this change won't be in 8.x either. Are we saying that in 8.1, the change is no longer needed because the root problem is known to have been solved in another way?

**Comment by grubbins** [comment]
If that's our only option then OK. In general I would expect some other,
new performance issues to come with a major version upgrade, as happened
with 6 and 7, so I would still prefer the 7.10 patch and waiting longer to
upgrade - but will leave it to our operational team to decide, and cease
grumbling :)

**Comment by angeloskyratzakos** [comment]
workspace fails to get update with error:
```
Error: failed to migrate with the plan: driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: duplicate key value violates unique constraint "db_migrations_pkey", query: INSERT INTO db_migrations (version, name) VALUES (110, 'update_vacuuming')
```

**Comment by pvev** [comment]
Closing this PR since no longer required.

**Pull request closed without merging**
